[unittests] Create OperationGradTest to isolate unit tests for gradient computation

d1jang opened the pull request:
*Description*:
Please see the comment for computeGradient().

*Testing*:
Run GradCheckTest

*Documentation*:

Partially fixing #2513

Please see a detailed explanation of how to fill out the fields in the relevant sections in PULL_REQUEST.md.

artemrakhov commented on tests/unittests/GradCheckTest.cpp:
```diff
@@ -168,36 +168,135 @@ void performGradCheck(ExecutionEngine &EE, PlaceholderBindings &bindings,
+  bindings.allocate(B)->getHandle() = {3, 4};
+
+  Node *O = F->createAdd("add", A, B);
+  // O = { { a1 + b1 }, { a2 + b2 } }
```
We have similar but different comment in the actual grad computing code:
https://github.com/pytorch/glow/blob/2793672d9368f47690208acc3c8933944136cef9/lib/Optimizer/Lower.cpp#L47

d1jang replied:
Thanks! I'll look into that and let me know if the style of this comment should be unified.

d1jang commented on tests/unittests/GradCheckTest.cpp:
```diff
@@ -168,36 +168,135 @@ void performGradCheck(ExecutionEngine &EE, PlaceholderBindings &bindings,
 }
 
-TEST_P(InterpreterGrad, gradientCheckConcat) {
+/// Compute gradient of inputVar based on the one of outputNode.
```
I updated the comment on my best understanding. Please let me know if the new comment looks less confusing.

d1jang commented:
I addressed the issues you mentioned. Created a new test suite, OperationGradTest, and created a unittest for FC. Please take another look at the most recent commit.

artemrakhov commented on tests/unittests/OperatorGradTest.cpp:
```diff
@@ -0,0 +1,137 @@
+  // dF/dW = [ [8*1,   11*1 ], [8*2,   11*2 ] ]
+  expected.getHandle() = {8 * 1, 11 * 1, 8 * 2, 11 * 2};
+  EXPECT_TRUE(expected.isEqual(*getGradTensor(varGrads, W)));
+}
```
Please add a TODO for grads of x and b

d1jang replied:
Done.

artemrakhov commented on tests/unittests/OperatorGradTest.cpp:
```diff
@@ -0,0 +1,137 @@
+  ///
+  /// \param outputNode Node that contains result of f(x).
+  VariableGradientsList computeVarGrads(Node *outputNode) {
+    // A RegressionNode is created just to generate a fake gradient with the
```
I suggest to remove this comment, because we already explain purpose of RegressionNode in comment above.

artemrakhov commented on tests/unittests/OperatorGradTest.cpp:
```diff
@@ -0,0 +1,137 @@
+    EE_.compile(CompilationMode::Train, recordNet);
+
+    // Train the network just once to record the values of gradient for
+    // inputVar.
```
inputVar -> all input variables.

artemrakhov commented on tests/unittests/OperatorGradTest.cpp:
```diff
@@ -0,0 +1,137 @@
+#include "glow/ExecutionEngine/ExecutionEngine.h"
+#include "glow/Graph/Graph.h"
+#include "glow/Graph/Nodes.h"
+#include "glow/IR/IR.h"
```
Why do you need to include IR/ ?

artemrakhov commented on tests/unittests/OperatorGradTest.cpp:
```diff
@@ -0,0 +1,136 @@
+  /// Compute gradients of placeholders in bindings_. Given
+  /// outputNode, representing H(Vars), this function will train H(vars) to be
+  /// 0. It is achieved by creating RegressionNode between outputNode and 0,
+  /// which will minimize F, a total squared error of H devided by 2.
```
devided -> divided

d1jang replied:
Done.
